BCPopover/BCPopover.m

```diff
@@ -138,23 +162,42 @@ - (NSRect)popoverWindowFrame {
 - (NSRect)screenAnchorRect {
   if (self.attachedToView) {
     NSView *anchoredView = self.attachedToView.superview;
-    return [self.attachedToView.window convertRectToScreen:[anchoredView convertRect:anchoredView.bounds toView:nil]];
+    NSRect rect = [anchoredView.window convertRectToScreen:[anchoredView convertRect:anchoredView.bounds toView:nil]];
+    return [self ensureRectFitsInParentWindow:rect];
   } else
     return [self.window frame];
 }
 
 - (NSPoint)attachToPointInScreenCoordinates {
   if (self.attachedToView) {
-    NSRect screenRect = NSInsetRect([self.attachedToView convertRect:[self.attachedToView bounds] toView:nil], -6, -6);
-    NSPoint pointAtEdge = [self pointAtEdge:self.preferredEdge ofRect:screenRect];
-    NSRect converted = [self.attachedToView.window convertRectToScreen:NSMakeRect(pointAtEdge.x, pointAtEdge.y, 0, 0)];
-    return converted.origin;
+    NSRect rectInWindow = NSInsetRect([self.attachedToView convertRect:[self.attachedToView bounds] toView:nil], -6, -6);
+    NSPoint pointAtEdge = [self pointAtEdge:self.preferredEdge ofRect:rectInWindow];
+    
+    NSWindow *window = self.attachedToView.window;
+    
+    NSRect converted = [window convertRectToScreen:NSMakeRect(pointAtEdge.x, pointAtEdge.y, 0, 0)];
+    return [self ensureRectFitsInParentWindow:converted].origin;
   } else {
     NSRect windowFrame = self.window.frame;
     return NSMakePoint(NSMinX(windowFrame), NSMaxY(windowFrame));
   }
 }
 
+- (NSRect)ensureRectFitsInParentWindow:(NSRect)rect; {
```

Comment:
Your "YAGNI" is my "ICBB" (see if you can work out what that stands for ;) )

The tests are there to protect against future mistakes in refactoring, not just to prove that the current code is correct.

Whether you actually change the behaviour is a different issue. If you're happy that it just does what it does, then all I'm suggesting is that you name it accordingly.